### ISSUE TITLE
iOS 17.2 Beta - `Cannot inject key into script value`

### ISSUE THREAD

**Issue by vojto** [firebase/firebase-js-sdk#7742]
### Operating System

iOS 17.2

### Browser Version

Safari/605.1.15

### Firebase SDK Version

10.5.2

### Firebase SDK Product:

Auth, Firestore

### Describe your project's tooling

React app with Firebase, wrapped in Capacitor

### Describe the problem

About 100 of our users (I'm from [Reflect](https://reflect.app)) upgraded from iOS 17.1 to iOS 17.2 Beta.

Right away they started reporting broken sync. I reproduced it by upgrading my own device, and I can confirm that sync stops working.

At the very beginning there's this error:

```
UnknownError: Cannot inject key into script value
```

Firebase doesn't print any more log messages, even though I set the log level to `debug`.

I don't have a reproduction yet - however here's a Loom which goes over the debugger session: https://www.loom.com/share/9a6e78af37a3448294158b86640924ef

It's not clear whether this would reproduce with any Firebase project, or if there's something specific about ours.

I also don't know yet how could I reproduce it over an over again. It's clear that Firebase's IndexedDB on 17.1 doesn't migrate properly to 17.2. For now, I'll avoid uninstalling the app, as that would mean I couldn't reproduce anymore.

### Steps and code to reproduce issue

1. Create Firebase app (auth, Firestore) and open it in Safari on iOS 17.1. Make sure offline persistence is enabled.
2. Upgrade the device to iOS 17.2. 
3. Open the app again. It will print the error above, and the sync will hang.

**Comment by maccman** [comment]
To be clear, we know that this is a Firebase indexed DB state issue because a reinstall of our app fixes it.

**Comment by MarkDuckworth** [comment]
We will start looking into it. Can you provide details about how your app is calling the Firestore APIs when you experience the error? Is this just in app initialization or have you made any queries with Firestore when this occurs?

**Comment by vojto** [comment]
> We will start looking into it. Can you provide details about how your app is calling the Firestore APIs when you experience the error? Is this just in app initialization or have you made any queries with Firestore when this occurs?

We use onSnapshot, and I think we don't even make it to that point - but I'll double check.

It looks like initialization error, since it's related to reading from IndexeDB.

**Comment by MarkDuckworth** [comment]
@vojto, can you verify my understanding that the only debug log message from the Firebase SDK is 
`UnknownError: Cannot inject key into script value`?

**Comment by vojto** [comment]
@MarkDuckworth 

This is all the output we get from Firebase SDK:

```
[2023-11-02T06:57:59.041Z]  @firebase/firestore:"Firestore (10.5.2): AsyncQueue""Visibility state changed to hidden"
[2023-11-02T06:58:00.335Z]  @firebase/firestore:"Firestore (10.5.2): FirebaseAppCheckTokenProvider""AppCheck not yet detected"
```

I set the log level to debug using:

```
firebase.setLogLevel('debug')
firestore.setLogLevel('debug')
```

Our code reaches `firestore.onSnapshot`, but the callback is never called.

**Comment by mikelehen** [comment]
👋 Found this bug via Google search and am chiming in in case it's helpful. I'm running into this same `Cannot inject key into script value` error message from MacOs users on Safari 17.2 Beta but in a different context (not Firebase). The error seems to be caused by the IndexedDb cursor encountering an internal key that can't be converted into a JavaScript value (possibly because it contains null segments?). Source code here: https://github.com/WebKit/WebKit/blob/5f0cd64bb733f5b27328ae404e23a7ea1c15d1aa/Source/WebCore/bindings/js/IDBBindingUtilities.cpp#L522

Looks like there was a recent change to that code (Sep 30): https://github.com/WebKit/WebKit/commit/1b624bea4bbbf7cf1c9a7b8fc65e9fa6d153d32c#diff-eddc28288bb0c31a5054f13eff49750547527e4097ad402e3316716054813aa8L359

But I'm struggling to see an obvious regression, so it's _possible_ it's just a coincidence.

**Comment by mustafa0x** [comment]
- I've seen this 14 times in the last 24 hours for 3 users.
- Not firebase; using https://github.com/jakearchibald/idb
- Only safari 17.2, on iphone and mac
- I think a safari bug needs to be filed
- Sentry screenshot:
![image](https://github.com/firebase/firebase-js-sdk/assets/1473102/87125e4a-18cc-408e-9cd6-4690329944d2)

**Comment by MarkDuckworth** [comment]
It seems like this is not a Firestore specific issue. We will keep an eye on it, but there's not much we can do about it now.

**Comment by mustafa0x** [comment]
I filed an issue with safari: https://bugs.webkit.org/show_bug.cgi?id=264099

No reproduction. This is the stack trace:

```
UnknownError
Cannot inject key into script value
/idb/build/wrap-idb-value.js in success at line 37:26

        const unlisten = () => {
            request.removeEventListener('success', success);
            request.removeEventListener('error', error);
        };
        const success = () => {
            resolve(wrap(request.result)); // ❗The line❗
            unlisten();
        };
        const error = () => {
            reject(request.error);
            unlisten();
```

**Comment by MarkDuckworth** [comment]
@vojto, I empathize with your situation and understand why you want a workaround. I spoke with a colleague and we didn't feel like it was wise to spend development cycles working around a regression in a beta version of Safari that's also affecting other software. Any cycles spent on this take away from other work in Firebase. I will ask for a second opinion though.

You have a lot of users already on 17.2, do you feel like your users represent general population or does your software target the development community?

Is it possible to temporarily workaround the issue in your software? Detect the affected browser and clear the SDK cache or prompt the user to clear browser cache / re-install?

**Comment by maccman** [comment]
@MarkDuckworth They're pretty late in the beta testing process. I guess the risk the Firebase team is running is that the beta gets released to production with the bug still in it. And then Firebase is suddenly broken across the entire mobile web. It might be better to get ahead of it.

**Comment by mustafa0x** [comment]
[Sihui Liu](https://bugs.webkit.org/show_bug.cgi?id=264099#c3):

> We are aware of this issue in iOS 17.2 beta and working on a fix. There is an issue in reading out IndexedDB items created on an old build, especially items created with key generator (https://w3c.github.io/IndexedDB/#key-generator-construct). 

> A potential workaround would be clearing those items when encountering this error.

**Comment by 1amageek** [comment]
<img width="715" alt="スクリーンショット 2023-11-07 17 08 43" src="https://github.com/firebase/firebase-js-sdk/assets/11146538/3d5be28e-ccec-4bd3-9712-1734ca9ca575">

**Comment by weisensee** [comment]
> @MarkDuckworth They're pretty late in the beta testing process. I guess the risk the Firebase team is running is that the beta gets released to production with the bug still in it. And then Firebase is suddenly broken across the entire mobile web. It might be better to get ahead of it.

@MarkDuckworth Just letting you know this happened. Our app is user facing, not for developers/technical people and the whole app is down on 17.2. Seems like this bug got automatically rolled out to production on iOS. We are working on a fix to blow out indexDB and see if it fixes it.

**Comment by vojto** [comment]
This is the snippet we added in [Reflect](https://reflect.app) if update is detected:

```
let store = WKWebsiteDataStore.default()
store.removeData(
    ofTypes: [WKWebsiteDataTypeIndexedDBDatabases],
    modifiedSince: Date(timeIntervalSince1970: 0),
    completionHandler: {}
)
```

It's been live for about a week, and seems to be working fine.

**Comment by tannerellen** [comment]
We are also experiencing this with our web app. I have had multiple users report this now and I have verified it happens by updating my phone to the iOS 17 beta. The Firebase Auth process errors out and my app is left at a white screen. This affects the latest iOS and MacOS betas. 

We are using the realtime database but the failure is in the auth module before accessing anything from the realtime database.

We have confirmed that deleting the Firebase indexDB entry resolves the issue but that isn't something we want to put in production as a work around if we can avoid it.

This is especially challenging because once the indexDB entry is cleared the issue no longer persists and is not reproducable (I would imagine unless you downgrade the OS and upgrade again). So this is a blocking bug that will affect everyone that updates their OS but is not easy to test as you can't enter a fixed / broken state easily.

**Comment by LandonS23** [comment]
Also experiencing this in our web app

**Comment by mehdi-tnt** [comment]
> This is the snippet we added in [Reflect](https://reflect.app) if update is detected. It's a capacitor app, so we can do in on the native side.
> 
> ```swift
> let store = WKWebsiteDataStore.default()
> store.removeData(
>     ofTypes: [WKWebsiteDataTypeIndexedDBDatabases],
>     modifiedSince: Date(timeIntervalSince1970: 0),
>     completionHandler: {}
> )
> ```
> 
> It's been live for about a week, and seems to be working fine.

Where did you add this code ? Which file ?

**Comment by tgangso** [comment]
> This is the snippet we added in [Reflect](https://reflect.app) if update is detected. It's a capacitor app, so we can do in on the native side.
> 
> ```swift
> let store = WKWebsiteDataStore.default()
> store.removeData(
>     ofTypes: [WKWebsiteDataTypeIndexedDBDatabases],
>     modifiedSince: Date(timeIntervalSince1970: 0),
>     completionHandler: {}
> )
> ```
> 
> It's been live for about a week, and seems to be working fine.

Would you mind sharing the full workaround? Including where you put it and version detection. Thanks

**Comment by vojto** [comment]
> Would you mind sharing the full workaround? Including where you put it and version detection. Thanks

Keep in mind this is for Capacitor apps:

There's a special Capacitor plugin for this, and it runs in the `load()` method. You can get the current version from `UIDevice.current.systemVersion`. In user defaults, keep the last seen system version, and then compare to check if it changed since last run.

The only problem is I cannot detect whether the version is currently a beta, or whether it's official release.

However I'm assuming one of the following betas will fix this problem and it won't be an issue anymore.

**Comment by tgangso** [comment]
> > Would you mind sharing the full workaround? Including where you put it and version detection. Thanks
> 
> Keep in mind this is for Capacitor apps:
> 
> There's a special Capacitor plugin for this, and it runs in the `load()` method. You can get the current version from `UIDevice.current.systemVersion`. In user defaults, keep the last seen system version, and then compare to check if it changed since last run.
> 
> The only problem is I cannot detect whether the version is currently a beta, or whether it's official release.
> 
> However I'm assuming one of the following betas will fix this problem and it won't be an issue anymore.

Yes, we are using capacitor. Thank you for the information. Are you able to share the full code to help save some time implementing the workaround?

**Comment by vojto** [comment]
> Yes, we are using capacitor. Thank you for the information. Are you able to share the full code to help save some time implementing the workaround?

Yes, I think that should be fine. Here you go: https://gist.github.com/vojto/509094a5327b6fc7983f5ebcef743b25

**Comment by tannerellen** [comment]
It looks like Apple fixed this on their end. Our app is no longer affected by this on the newest iOS 17.2 Beta 3 update. I’ll be doing more testing but as of now looks like this has been resolved.

**Comment by webmozart** [comment]
Confirmed fixed for our users! 🎉

**Comment by tgangso** [comment]
iOS 17.2 Beta 3 update seems to have fixed this here aswell.

**Comment by maccman** [comment]
We are now seeing an indexOf bug. We'll open a separate ticket for that.

**Comment by vojto** [comment]
> We are now seeing an indexOf bug. We'll open a separate ticket for that.

Here's the new ticket: https://github.com/firebase/firebase-js-sdk/issues/7806
